### ISSUE TITLE
Is it possible to only scan vulnerabilities for installed packages ?

### ISSUE THREAD

**Issue by dmeulen** [aquasecurity/trivy#993]
I use trivy to scan docker images in my gitlab pipelines. I recently discovered that trivy can also be a very useful tool to scan a linux virtual machine. I use a cronjob to scan the root file system once a day and expose the number of criticals and highs to prometheus via node_exporter. 

But I found that:
1. it is rather intrusive ( scanning the whole fs )
2. it uses way to much memory to be useful on small vm's ( prolly because of scanning the whole fs )

Is it possible to scan only the installed packages on the system without traversing the whole filesystem?
I tried skipping dirs, but this makes no difference in memory usage. I also tried scanning just a small part of the filesystem, but it can not determine the installed OS that way.

**Comment by tspearconquest** [comment]
Older versions did scan only packages, however it looks like aquasec have added non-os package scanning without a way to turn it off. Then they added the fs scanning feature which also uses that mandatory new feature.

I also would like to see a way to turn off non-OS package scanning for both "trivy image" and "trivy fs".

Trivy v0.18.1 is itself affected by 2 non-OS package vulnerabilities which we users cannot simply patch around in our Dockerfiles, and so this is a requirement in order for us to be able to deploy the latest version.

**Comment by dmeulen** [comment]
I am aware of the `--vuln-type os` flag. When using it, it still scans the whole filesystem when passing it :-(. 

I was hoping for something that:

1. detects operating system
2. get the installed packages for the system
3. scan the list of installed packages
4. done

Thanks for the note, but we run/most of our tools that are not installed by the OS in docker containers. These are scanned by trivy in our pipelines.

I'll try to dive into the code and maybe make it so ;-) ( If I can find some time to spare )...

**Comment by knqyf263** [comment]
@dmeulen Thanks for the report. You're right. Even when you specify `--skip-dirs` and `--vuln-type`, `fs` subcommand traverses the entire fs. There are some reasons to do that, but we're planning to improve `--skip-dirs` not to traverse the directory actually. I believe we can include it in the next version.

**Comment by fboender** [comment]
I'm running into a similar problem. For various reasons we can't always integrate Trivy into a build pipeline or scan a registry. So I was very excited when I found out Trivy can scan an entire filesystem on a host. That incredibly simplifies scanning many hosts for us, since we can just automate it all with Ansible. However, the memory usage makes this unfeasible.

For example, peak memory usage while scanning various of the same parts of a filesystem:

```
          Res          Virt Cmd
    5372.0 Mb     6403.7 Mb /opt/trivy/trivy -q fs /data 
    5026.7 Mb     6271.6 Mb /opt/trivy/trivy -q fs /data/cust/releases 
     617.0 Mb     1581.2 Mb /opt/trivy/trivy -q fs /data/cust/releases/release-2.2.0-build20210601
```

It seems Trivy keeps the entire fs tree structure meta data in memory, rather than freeing the memory after each subtree has been scanned? Otherwise I'd expect the scan of `/data/cust/releases` to not use much more memory than the largest subtree under `releases`. (note that there are only `release-X.X.X...` directories under `/data/cust/releases`, nothing else).

What's interesting is that Trivy does eventually start releasing memory again. Chronological samples from a single run when scanning the entire filesystem:

```
          Res          Virt Cmd
     109.1 Mb     1117.9 Mb root /opt/trivy/trivy -q fs / 
    2098.9 Mb     3100.4 Mb root /opt/trivy/trivy -q fs / 
    5005.8 Mb     6540.0 Mb root /opt/trivy/trivy -q fs / 
    5815.1 Mb     7663.1 Mb root /opt/trivy/trivy -q fs / 
    6028.8 Mb     8257.7 Mb root /opt/trivy/trivy -q fs / 
    5945.8 Mb     8257.7 Mb root /opt/trivy/trivy -q fs / 
    4924.2 Mb     8258.0 Mb root /opt/trivy/trivy -q fs / 
    4789.8 Mb     8258.0 Mb root /opt/trivy/trivy -q fs / 
    3870.6 Mb     8258.2 Mb root /opt/trivy/trivy -q fs / 
    2985.8 Mb     8258.2 Mb root /opt/trivy/trivy -q fs / 
    2889.9 Mb     8258.2 Mb root /opt/trivy/trivy -q fs / 
    2029.2 Mb     8266.2 Mb root /opt/trivy/trivy -q fs / 
```

Perhaps it's a garbage collection issue?

**Comment by fboender** [comment]
Out of curiosity, I tried some garbage collection things. Keep in mind that I know next to nothing about Go's GC and none of this may be relevant.

Running Trivy with `GODEBUG=gctrace=1` shows increasing wait periods between GCs (or just trace reports??) as Trivy approaches its maximum memory usage while scanning the entire filesystem. Once the period between GCs speeds up again, the memory usage also starts going down. This could be an artifact of the filesystem layout though.

Something else: If I run Trivy with `GOGC=10 /opt/trivy/trivy`, the maximum (resident) memory usage only goes up to 2.4 Gb, whereas without it goes to about 6 Gb. The runtime doesn't seem much longer. I did forget to clear the cache in between runs, but I did alternate runs of GOGC=100 and GOGC=10 three times with similar results on each run, so I'm guessing not clearing the cache doesn't really influence the results in this case.

So I guess it's better with GOGC=10, but 2.4 Gb is still a fair amount.
